widgetry/src/shaders/vertex_140.glsl

```diff
@@ -4,21 +4,39 @@
 uniform vec3 transform;
 // (window width, window height, z value)
 uniform vec3 window;
+// textures grid
+uniform sampler3D textures;
 
 layout (location = 0) in vec2 position;
-layout (location = 1) in vec4 style;
-out vec4 pass_style;
+layout (location = 1) in vec4 color;
+layout (location = 2) in float texture_index;
 
+out vec4 fs_color;
+out vec3 fs_texture_coord;
 void main() {
-    pass_style = style;
+    fs_color = color;
+
+    float zoom = transform[2];
 
     // This is map_to_screen
-    float screen_x = (position[0] * transform[2]) - transform[0];
-    float screen_y = (position[1] * transform[2]) - transform[1];
+    float screen_x = (position[0] * zoom) - transform[0];
+    float screen_y = (position[1] * zoom) - transform[1];
+
     // Translate that to normalized device coordinates (NDC)
     float x = (screen_x / window[0] * 2.0) - 1.0;
     float y = (screen_y / window[1] * 2.0) - 1.0;
 
     // Note the y inversion
     gl_Position = vec4(x, -y, window[2], 1.0);
+
+    // An arbitrary factor to scale the textures we're using.
+    //
+    // The proper value depends on the design of the particular sprite sheet.
+    // If we want to support multiple sprite sheets, this could become a 
+    // uniform, or a vertex attribute depending on how we expect it to change.
+    float texture_scale = 16.0;
+
+    float t_x = ((position[0] * zoom)) / texture_scale / zoom;
```

Comment:
I fixed the problem I had with textures "shifting" btw - the problem was that I was using the screen_x, which applies the transform. 

TBH I'm not sure I understand why it's not part of this math, only that "it looks right without it" and "it longs wrong with it".